src/backend/distributed/operations/shard_split.c

```diff
@@ -1145,3 +1236,707 @@ TryDropSplitShardsOnFailure(HTAB *mapOfShardToPlacementCreatedByWorkflow)
+																	  sourceShardToCopyNode
+																	  ->
+																	  workerPort,
+																	  superUser,
+																	  databaseName);
+	ClaimConnectionExclusively(sourceConnection);
+
+	HTAB *mapOfShardToPlacementCreatedByWorkflow =
+		CreateEmptyMapForShardsCreatedByWorkflow();
+
+	HTAB *mapOfDummyShardToPlacement = SetupHashMapForShardInfo();
+
+	/* Non-Blocking shard split workflow starts here */
+	PG_TRY();
+	{
+		/* 1) Physically create split children. */
+		CreateSplitShardsForShardGroup(mapOfShardToPlacementCreatedByWorkflow,
+									   shardGroupSplitIntervalListList,
+									   workersForPlacementList);
+
+		/*
+		 * 2) Create dummy shards due logical replication constraints.
+		 *    Refer to the comment section of 'CreateDummyShardsForShardGroup' for indepth
+		 *    information.
+		 */
+		CreateDummyShardsForShardGroup(
+			mapOfDummyShardToPlacement,
+			sourceColocatedShardIntervalList,
+			shardGroupSplitIntervalListList,
+			sourceShardToCopyNode,
+			workersForPlacementList);
+
+		CreateReplicaIdentities(mapOfDummyShardToPlacement,
+								shardGroupSplitIntervalListList, workersForPlacementList);
+
+
+		/* 3) Create Publications. */
+		CreateShardSplitPublications(sourceConnection, shardSplitHashMapForPublication);
+
+		/*
+		 * 4) Create template replication Slot. It returns a snapshot. The snapshot remains
+		 * valid till the lifetime of the session that creates it. The connection is closed
+		 * at the end of the workflow.
+		 */
+		MultiConnection *templateSlotConnection = NULL;
+		char *snapShotName = CreateTemplateReplicationSlotAndReturnSnapshot(
+			shardIntervalToSplit, sourceShardToCopyNode, &templateSlotConnection);
+
+
+		/* 5) Do snapshotted Copy */
+		DoSplitCopy(sourceShardToCopyNode, sourceColocatedShardIntervalList,
+					shardGroupSplitIntervalListList, workersForPlacementList,
+					snapShotName);
+
+		/* 6) Execute 'worker_split_shard_replication_setup UDF */
+		List *replicationSlotInfoList = ExecuteSplitShardReplicationSetupUDF(
+			sourceShardToCopyNode,
+			sourceColocatedShardIntervalList,
+			shardGroupSplitIntervalListList,
+			workersForPlacementList);
+
+		/*
+		 * Subscriber flow starts from here.
+		 * Populate 'ShardSplitSubscriberMetadata' for subscription management.
+		 */
+		List *shardSplitSubscribersMetadataList =
+			PopulateShardSplitSubscriptionsMetadataList(
+				shardSplitHashMapForPublication, replicationSlotInfoList);
+
+		/* Create connections to the target nodes. TODO: can be refactored */
+		List *targetNodeConnectionList = CreateTargetNodeConnectionsForShardSplit(
+			shardSplitSubscribersMetadataList,
+			connectionFlags,
+			superUser, databaseName);
+
+		/* 7) Create copies of template replication slot */
+		char *templateSlotName = ShardSplitTemplateReplicationSlotName(
+			shardIntervalToSplit->shardId);
+		CreateReplicationSlots(sourceConnection, templateSlotName,
+							   shardSplitSubscribersMetadataList);
```

Comment:
I think it would be good to move this to one of the earlier stages. Creating replication slots can fail for reasons like not having `max_replication_slots` set high enough. It's quite a bad user experience if this only errors after copying all the data. So I think it's nicer to open them as soon as possible.

Reply:
@JelteF : Given that we would fail in both the cases, won't it be a way to give 'second chance'? I mean while the copy was happening, based on the size it may take time. And when we create slots, the slots would have been available(potentially released by other consumers)?
Erroring out early seems fine too. I just had the above idea. 
Also in both the cases the slots have to be increased by user. If we error out early, he will do it before.

Reply:
I understand your reasoning for the "second chance", but I feel like that's covering a quite unlikely edge case where there are enough slots, but they are temporarily used by something else. I think it's much more likely that the maximum number of slots is simply misconfigured to be too low, and in that case it's really annoying for a user to get an error about after waiting for multiple hours for the data copy.

Reply:
In the case where the second chance would work the user experience is not made much worse by erroring early, because realistically the maximum number of replication slots is set too low to handle all replication that the user wants to do.

Reply:
@JelteF : Makes sense. Will try to reorder.